src/router.rs

```diff
@@ -5,27 +5,78 @@ use crate::parser::parse;
 use crate::schema::Schema;
 use crate::semantics::{FieldCounter, Validate};
 use std::collections::{BTreeMap, HashMap};
+use std::ops::Deref;
 use uuid::Uuid;
 
 #[derive(Debug, PartialEq, Eq, PartialOrd, Ord)]
 struct MatcherKey(usize, Uuid);
 
 #[derive(Debug)]
 pub struct Router<'a> {
-    schema: &'a Schema,
+    schema: SchemaOwnedOrRef<'a>,
     matchers: BTreeMap<MatcherKey, Expression>,
     pub fields: HashMap<String, usize>,
 }
 
+/// A smart pointer over a [`Schema`], which may be either borrowed or owned.
+///
+/// Used by [`Router`] to support both externally managed and self-contained schemas.
+/// Owning the schema is especially useful when the router is used outside of the FFI context,
+/// making it fully independent.
+///
+/// Implements [`Deref`] for ergonomic access to the underlying [`Schema`].
+#[derive(Debug)]
+enum SchemaOwnedOrRef<'a> {
+    Ref(&'a Schema),
+    Owned(Box<Schema>),
```

Comment:
an alternative here could be using an Arc so that the same `Arc<Schema>` can be shared across routers while using the `Owned` variant:
```suggestion
    Owned(Arc<Schema>),
```

---

I also considered using an `Arc` **instead** of the current approach (to replace the `SchemaOwnedOrRef` pointer entirely), but it seems this would require modifying the FFI interface, which is probably a larger scope than we want for this PR.

Looking for some feedback on this @dndx - thanks

Reply:
For example, `router_new` [here](https://github.com/Kong/atc-router/blob/afe0b973eb971e0b9dd34d99f05ec1cd3ff27353/src/ffi/router.rs#L27):

> pub unsafe extern "C" fn router_new(schema: &Schema) -> *mut Router 

If we were to change our Router struct to something like:
```rust
pub struct Router {
    schema: Arc<Schema>,
    matchers: BTreeMap<MatcherKey, Expression>,
    pub fields: HashMap<String, usize>,
}
```

Then in order to maintain the ability to share a schema across routers we should have:

```rust
pub fn new(schema: Arc<Schema>) -> Self {
    Self {
        schema,
        matchers: BTreeMap::new(),
        fields: HashMap::new(),
    }
}
```

but that would be incompatible with the current schema parameter passed to `router_new` as a reference (which I think we want to maintain unchanged, at least for the scope of this PR).

Reply:
Oh, I thought we were just changing `enum SchemaOwnedOrRef` to use `Arc` instead of `Box`, and that will not affect the previous reference-based interface. For changing `Router.schema`, it indeed introduces a breaking change to our FFI, I'm not sure whether this is feasible.

Reply:
yes, sorry, let me rephrase that.

What I meant is that the current change (using SchemaOwnedOrRef) is safe, changing the owned Schema to use Arc or Box is also safe and we could decide to do that easily (but based on another [closed discussion](https://github.com/Kong/atc-router/pull/318#discussion_r2192414371) above we are probably not going to anyway).

An alternative approach where, instead of using `SchemaOwnedOrRef`, we just replace the current Schema reference with an `Arc` would require an FFI interface change as explained above, and I agree with you that it is probably not feasible, so I think we are on the same page.